### PR TITLE
feat(adapter): Improve Prisma Adapter

### PR DESCRIPTION
**Pull request by williamluke4**
- [x] Add `lru-cache`
- [x] Make the schema more "Prismic"

### Diff between old schema and new schema
```diff
 model Account {
-  id                 Int       @default(autoincrement()) @id
-  compoundId         String    @unique @map(name: "compound_id")
-  userId             Int       @map(name: "user_id")
-  providerType       String    @map(name: "provider_type")
-  providerId         String    @map(name: "provider_id")
-  providerAccountId  String    @map(name: "provider_account_id")
-  refreshToken       String?   @map(name: "refresh_token")
-  accessToken        String?   @map(name: "access_token")
-  accessTokenExpires DateTime? @map(name: "access_token_expires")
-  createdAt          DateTime  @default(now()) @map(name: "created_at")
-  updatedAt          DateTime  @default(now()) @map(name: "updated_at")
-
-  @@index([providerAccountId], name: "providerAccountId")
-  @@index([providerId], name: "providerId")
-  @@index([userId], name: "userId")
-
-  @@map(name: "accounts")
+  id                 Int       @id @default(autoincrement())
+  userId             Int
+  providerType       String
+  providerId         String
+  providerAccountId  String
+  refreshToken       String?
+  accessToken        String?
+  accessTokenExpires DateTime?
+  createdAt          DateTime  @default(now())
+  updatedAt          DateTime  @updatedAt
+  user               User      @relation(fields: [userId], references: [id])
+
+  @@unique([providerId, providerAccountId], name: "providerId_providerAccountId_unique")
 }
 
 model Session {
-  id           Int      @default(autoincrement()) @id
-  userId       Int      @map(name: "user_id")
+  id           Int      @id @default(autoincrement())
+  userId       Int
   expires      DateTime
-  sessionToken String   @unique @map(name: "session_token")
-  accessToken  String   @unique @map(name: "access_token")
-  createdAt    DateTime @default(now()) @map(name: "created_at")
-  updatedAt    DateTime @default(now()) @map(name: "updated_at")
-
-  @@map(name: "sessions")
+  sessionToken String   @unique
+  accessToken  String   @unique
+  createdAt    DateTime @default(now())
+  updatedAt    DateTime @updatedAt
+  user         User     @relation(fields: [userId], references: [id])
 }
 
 model User {
-  id            Int       @default(autoincrement()) @id
+  id            Int       @id @default(autoincrement())
   name          String?
   email         String?   @unique
-  emailVerified DateTime? @map(name: "email_verified")
+  emailVerified DateTime?
   image         String?
-  createdAt     DateTime  @default(now()) @map(name: "created_at")
-  updatedAt     DateTime  @default(now()) @map(name: "updated_at")
-
-  @@map(name: "users")
+  createdAt     DateTime  @default(now())
+  updatedAt     DateTime  @updatedAt
+  accounts      Account[]
+  sessions      Session[]
 }
 
 model VerificationRequest {
-  id         Int      @default(autoincrement()) @id
+  id         Int      @id @default(autoincrement())
   identifier String
   token      String   @unique
   expires    DateTime
-  createdAt  DateTime  @default(now()) @map(name: "created_at")
-  updatedAt  DateTime  @default(now()) @map(name: "updated_at")
-
-  @@map(name: "verification_requests")
+  createdAt  DateTime @default(now())
+  updatedAt  DateTime @updatedAt
 }
+
 ```
```

### CODE REVIEW

**Comment by BjoernRave** [comment]
what is blocking from this getting merged?

**Comment by balazsorban44** [comment]
Hi @BjoernRave, first of all there are some merge conflicts that must be fixed before merge. Also, someone from the core team should review this. I am unfortunately only familiar with the database less flows, so it is waiting for someone else.

I for one cannot even decide if this is a breaking change, or not.

**Comment by williamluke4** [comment]
@balazsorban44 Merge conflicts have been resolved

File: package.json
Comment (balazsorban44) [inline package.json:50]: `jwt-decode` has been removed, in favor of `jsonwebtoken`, it should be removed here as well.

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Any way of making the two new dependencies optional? I don't think these should be downloaded for those who do not use the Prisma Adapter at all.

We recently opened up a repository for new adapters at https://github.com/nextauthjs/adapters

I am wondering if this `PrismaNew` could be the first official adapter there, leading the way for the community with a good example. 🙂

We plan to separate out adapters and not adding more to the core anyway, exactly because of the additional packages they bring to users who do not need them, and we wish them to be maintainable directly by the community.

As of now, we have TypeORM and Prisma built-in, and even when we move these to another place, they can be re-imported into the core to not cause any breaking change from the user perspective.

It would just be easier to maintain, I think.

Also adding additional required packages for certain adapters would be more transparent and something the users could be made aware.

If you agree, could you please move these changes to https://github.com/nextauthjs/adapters in a PR wrapped as an npm package, with the name `@next-auth/prisma-adapter`?

If you wish to maintain this adapter in the future, I can also add you as a maintainer to the repository.

**Comment by williamluke4** [comment]
@balazsorban44 I'll have a chat with the team internally then I'll get back to you

**Comment by williamluke4** [comment]
Closing as work on this has moved to https://github.com/nextauthjs/adapters/pull/1

**Pull request closed without merging**
